1단계 - 스프링 프레임워크 적용

hjhello423 opened the pull request:
안녕하세요 :)
1단계 미션 진행했습니다!
리뷰 잘 부탁드립니다! 🙇‍♂️

hjhello423 committed:
docs: 요구사항 정리

hjhello423 committed:
feat: controller 생성, 테스트 코드 추가

hjhello423 committed:
feat: request dto, 테스트 코드 추가

hjhello423 committed:
feat: RacingCar 도메인, 테스트 코드 추가

hjhello423 committed:
feat: 이동 전략 추가

hjhello423 committed:
feat: racingcar 일급객체, 테스트 코드 추가

hjhello423 committed:
feat: 우승자를 가리는 기능 추가, 테스트 추가

hjhello423 committed:
test: 테스트용 이동 전략 추가

hjhello423 committed:
feat: 게임 관리 도메인 추가

hjhello423 committed:
feat: 게임 플레이 service 비즈니스 개발, response 정의

hjhello423 committed:
feat: 컨트롤러에 서비스 반영, 테스트 코드 수정

hjhello423 committed:
docs: 개발 진행 상황을 요구 사항 리스트에 반영

hjhello423 committed:
docs: 테이블 추가 및 수정

hjhello423 committed:
chore: h2 db 관련 의존성, 설정 추가

hjhello423 committed:
feat: 게임 결과를 db에 저장하는 기능 추가

hjhello423 committed:
feat: service에 db에 저장하도록 변경

hjhello423 committed:
docs: DB 관련 요구 사항 리스트에 반영

hjhello423 committed:
refactor: 숫자가 4이상일 경우에만 차가 전진하도록 기능 수정, 테스트 코드 수정

hjhello423 committed:
fix: 커밋 꼬인 부분 반영

hjhello423 committed:
refactor: 상수명 수정

hjhello423 committed:
docs: 기능 개발 사항 반영

hjhello423 committed:
fix: 오탈자

ecsimsw commented on src/main/java/racingcar/domain/RacingCars.java:
```diff
@@ -0,0 +1,44 @@
+
+public class RacingCars {
+
+    private List<RacingCar> racingCars;
```
racingCars에 값 변경을 막을 수 있도록 final을 추가하면 어떨까요?

hjhello423 replied:
final을 잘 안 붙이는 게 습관이 되어있어서 자꾸 놓치네요🥲🥲
리뷰 감사합니다! 다른 곳도 final을 추가해 보겠습니다!

ecsimsw replied:
제가 열심히 짚어보겠습니다 👀👀

ecsimsw commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,28 @@
+
+    public String getWinners() {
+        List<String> carNames = racingCars.getMaxDistanceCarName();
+        return String.join(",", carNames);
```
RacingGame이 반환하는 우승자 형태가 좀 더 유연하면 어떨까요?
RacingGame은 딱 자동차를 count만큼 이동하는데만 신경을 쓰고, 우승자 자동차들이 어떤 형태("A,B,C")로 저장되는지 또는 출력되는 모르도록 하면 더 좋을 것 같다는 생각이 드는데 어떻게 생각하시나요? 😊

hjhello423 replied:
`getWinners()`이 처음에는 `List<String>`을 반환하도록 개발을 했었는데요.
API에 응답하는 부분과, DB에 insert 하는 부분에서 List 형태로 반환하다 보니 `[a, b, c]`처럼 출력 되더라고요.
우승자를 출력하는 곳이 API 응답에 만 있었다면 request dto 같은 곳에서 처리를 고민해 봤을 텐데, DB에도 해당 값을 동일하게 넣어야 하다 보니 각각의 영역에서 출력 결과를 변환하는 메서드를 만들기보다는 RacingGame 도메인으로 옮겨보았습니다!

우선 List 타입으로 반환하도록 코드를 다시 수정해 보았는데요.
이 부분에 대한 진환님의 의견도 궁금하네요😀

ecsimsw replied:
아하!! 그렇다면 PlayHistory라는 엔티티를 따로 만들어 관리해보면 어떨까요? 
RacingGame은 DB에 저장꼴에 관여하지 않도록 하고 PlayHistoryRepository는 RacingCars가 아닌 PlayHistory 를 다루도록요.

hjhello423 replied:
좋은 의견 주셔서 반영해 보았습니다! 👍 https://github.com/next-step/spring-racingcar/pull/49/commits/12923113e58c196ac9c912ad44db997565a18484

ecsimsw commented on src/main/java/racingcar/domain/RandomMoveStrategy.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public int move() {
+        Random random = new Random();
```
Random 객체를 움직일 때마다 매번 생성되어야 하면 비효율적인 것 같아요.
다른 방식으로 Random 인스턴스를 관리해볼까요?

ecsimsw commented on src/main/java/racingcar/infra/PlayHistoryDao.java:
```diff
@@ -0,0 +1,45 @@
+
+    private static final String TABLE_NAME = "player_history";
+    private static final String ID = "id";
+    private static final String[] COLUMNS = {"name", "position"};
```
오.. 하드 코딩을 빼시면서도 네이밍이 간결하고 명확해서 아래 코드가 훨씬 잘 읽혔습니다. 👍👍👍

ecsimsw commented on src/main/java/racingcar/presentation/dto/PlayRequest.java:
```diff
@@ -0,0 +1,33 @@
+        this.count = count;
+    }
+
+    @JsonIgnore
```
👍

ecsimsw commented on src/main/java/racingcar/presentation/dto/PlayResponse.java:
```diff
@@ -0,0 +1,57 @@
+
+    private static class RacingCarDto {
+        private String name;
+        private int position;
```
깔끔하네요!!!
여기도 필드에 final 적용될 수 있을 것 같아요.

다른 것보다도 다른 파일에선 final 처리가 된 곳이 있어 아예 통일되면 더 좋을 것 같아 말씀드려요 😊

hjhello423 replied:
https://github.com/next-step/spring-racingcar/pull/49#discussion_r1179286988 코멘트를 보고 다른 곳들도 수정해 보았습니다!

ecsimsw commented on src/main/java/racingcar/infra/PlayHistoryDao.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public void save(RacingCars racingCars) {
+        MapSqlParameterSource[] paramsList = createParam(racingCars);
```
이 정도는 기호차로 넘어가려고 했다가 createParam()의 반환 값을 받는 변수명을 보고 paramsList가 헷갈릴 수 도 있겠다는 생각을 했습니다.
배열을 반환하니 메서드명도 복수를 표시하면 어떨까요? 개인적으론 변수명도 params와 List가 복수 의미가 중복되니 params만으로도 충분히 의미가 전달될 것 같다는 생각도 드네요. (요건 패스하셔도 좋습니다 ㅎㅎ)

hjhello423 replied:
동의동의!
자동완성으로 만들어지면서 변수명에 list를 넣었네요. 자주 놓치는 부분인데 리뷰 감사합니다! 👍

ecsimsw replied:
이 부분도 final 처럼 일부러라도 더 열심히 봐보겠습니다 ㅎㅎ
습관 만들기에 제가 도움이 되셨으면 좋겠네요.

ecsimsw commented on src/main/java/racingcar/domain/RacingCars.java:
```diff
@@ -0,0 +1,44 @@
+
+    public List<RacingCar> getRacingCars() {
+        return racingCars;
+    }
```
전 이게 좀 헷갈렸어요!! RacingCars 의 인스턴스 변수명이 racingCars이니 헷갈리더라고요.
특히 이 파일 안에서는 문제가 없는데 RacingCars 타입의 변수명이 racingCars면서 그 안에 racingCars.racingCars를 꺼내게 되면... 

```
private MapSqlParameterSource[] createParam(RacingCars racingCars) {
       List<RacingCar> cars = racingCars.getRacingCars();
}
```
딱 예시로 필요했던 상황이 다른 파일에 있어 가져왔어요.
흠.. 딱 예쁜 꼴의 네이밍이 없을까요?  🤔
홍준님도 공감하신다면 함께 고민봐요 ㅎㅎ

hjhello423 replied:
이 부분은 value로 수정해 보았어요.
어떻게 생각하시나요?

ecsimsw commented on src/test/java/racingcar/presentation/GameControllerTest.java:
```diff
@@ -0,0 +1,46 @@
+        // then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(response.jsonPath().getString("winners")).isNotNull();
+        assertThat(response.jsonPath().getList("racingCars").size()).isEqualTo(3);
```
3을 변수로 빼서 역할(playCount)를 설명함과 동시에 하드 코딩을 제거하는 역할로 사용하면 어떨까요?

ecsimsw commented on src/test/java/racingcar/fixture/TestMoveStrategy.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public int move() {
+        return distance;
+    }
```
전략 패턴을 잘 사용하셨네요!! 
테스트 용이를 위한 전략 패턴. 최고십니다 👏

ecsimsw commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,10 @@
+      enabled: true
+  datasource:
+    url: jdbc:h2:mem:testdb;MODE=MySQL
+    driver-class-name: org.h2.Driver
```
오호!! properties보다 yaml을 더 선호하시는군요 😄

hjhello423 replied:
아무래도 들여 쓰기로 구분되다 보니까 훨씬 눈에 잘 들어오더라고 중복 단어도 많이 없어지더라고요.
그래서 개인적으로는 yml을 선호합니다! 😀

ecsimsw reviewed (approved):
안녕하세요. 이번 미션 리뷰를 맡게 된 김진환입니다.
반갑습니다.

빠르게 너무 잘하셨네요!!!
네이밍이나 구조, 테스트 코드, 포맷까지 너무 꼼꼼하게 잘해주셨습니다.

특별히 문제되는 부분은 없고, 같이 봤으면 좋겠는 것들 몇 개 커멘트 남겨봤으니 확인 부탁드려요.
고생하셨습니다 👍

hjhello423 committed:
refactor: 리뷰 반영

- final 적
- RacingCars의 racingCars 필드 이름 수정
- RacingGame의 우승자 출력 메서드 수정
- Random static 처리

hjhello423 committed:
test: 테스트 코드의 매직 리터럴에 대한 리뷰 반영

hjhello423 committed:
refactor: db 저장 로직 수정, 엔티티 추가
